rules/firestore.json

```diff
@@ -45,40 +51,132 @@ service cloud.firestore {
-            // Anybody can create a user
-            allow create: if true;
-            // Only a user can update its own fields
-            allow update: if isOwner(userId);
-            // Anybody can read the fields of a user
-            allow read: if true;
+        function isOwner(userId) {
+            return request != null &&
+                request.auth != null &&
+                request.auth.uid == userId;
         }
-        match /parties/{id} {
-            allow write, read: if userVerified();
+        function doesNotModifyUsernameIfSet() {
+            let usernameNotYetSet = hasField(resource.data, 'username') == false;
+            let usernameBeingSet = hasField(request.resource.data, 'username');
+            // It sometimes happens that firebase sends the username as part as updates, even though it is not actually updated. Hence, we accept usernames as part of updates, only if it doesn't change.
+            let doesNotChange = hasField(resource.data, 'username') && hasField(request.resource.data, 'username') && resource.data['username'] == request.resource.data['username'];
+            return usernameNotYetSet || usernameBeingSet == false || doesNotChange;
         }
-        match /joiners/{id} {
-            allow write, read: if userVerified();
+        function onlyModifiesVerifiedIfVerified() {
+            return hasField(request.resource.data, 'verified') == false ||
+                   request.resource.data.verified == false ||
+                   (userIsVerified() && resource.data.username != null);
         }
-        function userVerified() {
-            return isAuthentified() &&
+        match /parties/{id} {
+            // Anybody can create a part
+            allow create: if userIsVerified();
+            // The owner of a part can delete it
+            allow delete: if userIsVerified() &&
+                             creatorOf(id).id == request.auth.uid;
+            // Anyone can delete a part if the owner is not observing it or has timed out
+            allow delete: if userIsVerified() &&
+                             notObservingOrTimedOut(creatorOf(id), id);
+            // A number of fields are readonly:
+            // - typeGame
+            // - playerZero/playerOne once the part has started
+            // - beginning once the part has started
+            // the part is considered to be started as soon as beginning is set
+            // For any other fields, both players can modify them
+            allow update: if userIsVerified() &&
+                             preservesField('typeGame') &&
+                             (preservesField('playerZero') || resource.data.beginning == null) &&
+                             (preservesField('playerOne') || resource.data.beginning == null) &&
+                             (preservesField('beginning') || resource.data.beginning == null) &&
+                             (resource.data.playerZero == usernameOf(request.auth.uid) ||
```

Comment:
"tu crée pas de partie pour tes potes mon con ! Tu crée, tu joues!"
On peut rajouter si besoin qu'on peut pas dire "viens jouer à fou!" à quelqu'un qui observe pas la partie (sinon tu te retrouverais à perdre des parties dont t'as jamais vu la couleur haha :D )

Reply:
c'qui permettrait d'ailleurs un joli refactor de style "playersAreObservingThePart"

Reply:
désolé je comprends pas le lien entre ton commentaire et le code commenté :D

Reply:
Pour les modifications de playerZero: il me semble qu'à la création le créateur se met dedans.
Ensuite, je crois qu'à la création de la partie (après la dernière modif du joiner) le créateur pourra dire "je met le nom du chosenCandidate de la joiner dans playerZero et moi en playerOne ou l'inverse, mais que tout le reste sur playerZero n'est pas authorisé.

Donc un peu plus stricte que décrit n'est-ce-nain ?

Reply:
Mai, juin, juillet, rrroooh, et il serait pas invraissemblable d'avoir une réponse en aout haha :D !

Reply:
me semble que la logique playerZero/playerOne est correcte maintenant ? À tester la prochaine fois, une partie avec l'opposant en premier joueur

Reply:
Les tests que le flow normal marchent ne prouvent pas que la sécurité n'est pas plus stricte héhé :p !

Reply:
Ajoutés: 
  - should forbid creating a part with the wrong playerZero
  - should forbid starting a part when setting the non-chosenOpponent/non-creator playerZero
  - should forbid starting a part when setting the non-chosenOpponent/non-creator playerOne
(Et adapté les rules pour que ça soit vraiment le cas :p)

Reply:
je pige pas les formulation des deux dernières, comme l'indique des commentaires ailleurs je pense !

Reply:
corrigé !